### ISSUE TITLE
Compared with the native Torch cross-entropy, the gradient differences of the classifier are very large.

### ISSUE THREAD

**Issue by BangguWu** [apple/ml-cross-entropy#14]
I have implement a toy code 
```
import torch
import copy

from cut_cross_entropy import linear_cross_entropy
torch.manual_seed(42)

model = torch.nn.Linear(128, 128, dtype=torch.bfloat16).cuda()
classifier_weights = torch.nn.Linear(128, 32, dtype=torch.bfloat16).cuda()

model_c = copy.deepcopy(model)
classifier_weights_c = copy.deepcopy(classifier_weights)

input = torch.randn(8192, 128, dtype=torch.bfloat16).cuda()
input.requires_grad = True
input.retain_grad()

input_c = input.clone()

labels = torch.randint(0, 32, (8192,)).cuda()
labels_c = labels.clone()

embeddings = model(input)

shift_embeddings = embeddings[:-1, :]
shift_labels = labels[1:]

manual_shift_loss = linear_cross_entropy(shift_embeddings, classifier_weights.weight, shift_labels, impl="torch_compile")
manual_shift_loss.backward()

embeddings_c = model_c(input_c)
logits = classifier_weights_c(embeddings_c)
shift_logits_c = logits[:-1, :]
shift_labels_c = labels_c[1:]
manual_shift_loss_c = torch.nn.functional.cross_entropy(shift_logits_c, shift_labels_c)
manual_shift_loss_c.backward()

print(f"classifier_weights grad: {classifier_weights.grad}")
print(f"classifier_weights_c grad: {classifier_weights_c.grad}")
```

and the output is:

**Comment by zhixuan-lin** [comment]
Probably related to this, I observe that the loss increases during training using `linear_cross_entropy` with long-context training, while some other non-fused implementation works fine. So there is very likely something wrong with the implementation of `linear_cross_entropy` (probably numerical precision).

**Comment by zhixuan-lin** [comment]
My guess is `dc` ([here](https://github.com/apple/ml-cross-entropy/blob/1f3ebdb20653a26598a2722acd21e9b1528608c3/cut_cross_entropy/cce_backward.py#L269), the accumulator for the gradients of the linear layer weights) is kept in low precision. So when the sequence is long there could be huge numerical errors. This is also mentioned in the flash-linear-attention repository (see this [commit](https://github.com/sustcsonglin/flash-linear-attention/commit/30978aafe2339ca66960a89757f4d6d47412f0f6)).

**Comment by sudy-super** [comment]
Splitting it this way for gradient computation resulted in a slightly lower loss.
Could this be considered a fundamental solution?
```diff
 def cce_backward_kernel(
     do: torch.Tensor,
     e: torch.Tensor,
     c: torch.Tensor,
     lse: torch.Tensor,
     valids: torch.Tensor | None,
     softcap: float | None,
     filter_eps: float | None,
     targets: torch.Tensor | None = None,
     shift: bool = False,
     vocab_ordering: torch.Tensor | None = None,
     grad_scale: float = 1.0,
 ) -> tuple[torch.Tensor, torch.Tensor]:
     assert do.numel() in (e.size(0), 1)
     assert c.size(1) == e.size(1)
     assert lse.size(0) == e.size(0) or (valids is not None and lse.size(0) == valids.size(0))
     assert e.dtype in (
         torch.float16,
         torch.bfloat16,
     ), "Backwards requires embeddings to be bf16 or fp16"
     assert c.dtype in (
        torch.float16,
         torch.bfloat16,
     ), "Backwards requires classifier to be bf16 or fp16"
 
     do = do.contiguous()
     lse = lse.contiguous()

+    num_chunks = max(1, e.size(0) // 1024)
+    chunk_size = (e.size(0) + num_chunks - 1) // num_chunks
     de = torch.zeros_like(e)
     dc = torch.zeros_like(c)
+    for chunk_idx in range(num_chunks):
+        start_idx = chunk_idx * chunk_size
+        end_idx = min((chunk_idx + 1) * chunk_size, e.size(0))
+        
+        if end_idx <= start_idx:
+            continue
     assert de.stride() == e.stride()
     assert dc.stride() == c.stride()
```

**Comment by erikwijmans** [comment]
This is something we've noticed too, specifically of training models from scratch. 

@BangguWu is the loss gap on train or val? I have seldom seen train loss be different (except in the case of triton bugs that we haven't worked around yet), but I have seen val loss be different when training from scratch and the validation set has tokens that aren't present in the train set.

**Comment by erikwijmans** [comment]
We have been working on some updates on this [branch](https://github.com/apple/ml-cross-entropy/tree/ewijmans/kahan-updates). You can it via `pip install "cut-cross-entropy @ git+https://github.com/apple/ml-cross-entropy.git@ewijmans/kahan-updates"`. 

That adds two options to `linear_cross_entropy` and `patch`.

1. An `impl="cce_exact"` option that should match PyTorch exactly. That implementation is slower as it disables gradient filtering, but in our testing it isn't significantly slower in full model training and can be faster due the ability to increase the batch size.
2. A `use_kahan` flag that uses Kahan summation to deal with `dc` and `de` being kept in lower precision. This uses a bit more memory for temporary buffers. Setting `impl="cce_exact"` will automatically set `use_kahan=True`.

@BangguWu @zhixuan-lin If you are up for some beta testing, feel free to try these out and let me know how it goes.

**Comment by li-yi-dong** [comment]
Hi @erikwijmans 

Why not just use fp32 for `dc` and `de` for `_mm_backward`? And cast back to what precision `dc` and `de` suppose to be before return. The memory consumption seems to be the same as `use_kahan=True`.

**Comment by erikwijmans** [comment]
Maybe that is fixed in a newer versions of triton, but when I tried doing that triton would get thrown into an infinite loop. I'll see if that's fixed in newer versions and do fp32 summation when possible.

**Comment by erikwijmans** [comment]
Yep, looks like that bug was fixed in Triton 3.2. I'll update the code to use fp32 when possible.

**Comment by li-yi-dong** [comment]
> Yep, looks like that bug was fixed in Triton 3.2. I'll update the code to use fp32 when possible.

I tried to change the dc and de into fp32. It worked fine, without too much cost in time and space. And  in my training, using fp32 for accumulation of dc and de significantly improves stability.

**Comment by alxndrTL** [comment]
> Yep, looks like that bug was fixed in Triton 3.2. I'll update the code to use fp32 when possible.

Hello @erikwijmans, I pulled the latest version of `cce`, is passing `impl="cce_exact"` the right way to achieve what you describe here ?
I didn't have to upgrade to Triton 3.2.

(thank you for the great work btw)

**Comment by erikwijmans** [comment]
Yep, passing `impl="cce_exact"` should match PyTorch in all conditions. There is no need to upgrade to triton 3.2. The code will use fp32 when possible and kahan summation when not.

**Comment by erikwijmans** [comment]
@alxndrTL Actually, I think I figured it out. I had been assuming the TP library would want the embedding gradient to be just the portion for that rank and would synchronize those itself, but upon further review I don't think that is true. I think it expects CCE to synchronize that gradient. I made #42 that does that. If possible, can you install it via 

```
pip install "cut-cross-entropy @ git+https://github.com/apple/ml-cross-entropy.git@ewijmans/vocab-parallel-fix"
```

and try it out?

**Comment by alxndrTL** [comment]
Hello, indeed it solved the problem!
See the loss attached :
![Image](https://github.com/user-attachments/assets/cc13975d-649d-4f9b-80b9-033c3d637585)

Thank you!

**Comment by erikwijmans** [comment]
Glad to hear! I will merge that PR
